control/discovery.py

```diff
@@ -0,0 +1,1077 @@
+        nvme_keep_alive_dword15 = struct.unpack_from('<I', data, 68)[0]
+
+        pdu_reply = Pdu()
+        pdu_reply.type = NVME_TCP_PDU.RSP
+        pdu_reply.header_length = 24
+        pdu_reply.packet_length = 24
+
+         # Cqe for keep alive
+        keep_alive_reply = CqeNVMe()
+        keep_alive_reply.sq_head_ptr = self_conn.sq_head_ptr
+        keep_alive_reply.cmd_id = cmd_id
+
+        try:
+            conn.sendall(pdu_reply + keep_alive_reply)
+        except BrokenPipeError:
+            self.logger.error("client disconnected unexpectedly.")
+            return -1
+        self.logger.debug("reply keep alive request.")
+        return 0
+
+    def store_async(self, conn, data, cmd_id):
+        """Parse and store async event."""
+
+        self.logger.debug("parse and store async event.")
+        self_conn = self.conn_vals[conn.fileno()]
+        nvme_sgl = struct.unpack_from('<16B', data, 32)
+        nvme_sgl_desc_type = nvme_sgl[15] & 0xF0
+        nvme_sgl_desc_sub_type = nvme_sgl[15] & 0x0F
+        nvme_async_dword10 = struct.unpack_from('<I', data, 48)[0]
+        nvme_async_dword11 = struct.unpack_from('<I', data, 52)[0]
+        nvme_async_dword12 = struct.unpack_from('<I', data, 56)[0]
+        nvme_async_dword13 = struct.unpack_from('<I', data, 60)[0]
+        nvme_async_dword14 = struct.unpack_from('<I', data, 64)[0]
+        nvme_async_dword15 = struct.unpack_from('<I', data, 68)[0]
+
+        self_conn.recv_async = True
+        self_conn.async_cmd_id = cmd_id
+
+    def _state_notify_update(self, update, is_add_req):
+        """Notify and reply async event."""
+
+        for key in list(self.conn_vals.keys()):
+            if self.conn_vals[key].recv_async is True:
+                async_reply = CqeNVMe()
+                # async_event_type:0x2 async_event_info:0xf0 log_page_identifier:0x70
+                async_reply.dword0 = b'\x02\xf0\x70\x00'
+                async_reply.sq_head_ptr = self.conn_vals[key].sq_head_ptr
+                async_reply.cmd_id = self.conn_vals[key].async_cmd_id
+
+                try:
+                    self.conn_vals[key].connection.sendall(async_reply)
+                except BrokenPipeError:
+                    self.logger.error("client disconnected unexpectedly.")
+                    return -1
+                self.logger.debug("notify and reply async request.")
+                self.conn_vals[key].recv_async = False
+                return 0
+
+    def handle_timeout(self):
+        """Handle connection timeout."""
+
+        while True:
+            for key in list(self.conn_vals.keys()):
+                if self.conn_vals[key].keep_alive_timeout != 0 and \
+                  time.time() - self.conn_vals[key].keep_alive_time >= \
+                  self.conn_vals[key].keep_alive_timeout / 1000:
+                    # Adding locks to prevent another thread from processing sudden requests.
+                    # Is there a better way?
```

Comment:
See [this comment](https://github.com/ceph/ceph-nvmeof/pull/127#issuecomment-1696885405)